cloud/docker_client.go

```diff
@@ -97,6 +99,37 @@ func (c *dockerClientImpl) Init(apiVersion string) error {
 	return nil
 }
 
+// EnsureImageDownloaded checks if the image in s3 specified by the URL already exists,
+// and if not, creates a new image from the remote tarball.
+func (c *dockerClientImpl) EnsureImageDownloaded(ctx context.Context, h *host.Host, url string) (string, error) {
+	dockerClient, err := c.generateClient(h)
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to generate docker client")
+	}
+
+	// extract image name from url
+	baseName := path.Base(url)
+	imageName := strings.TrimSuffix(baseName, filepath.Ext(baseName))
+
+	// check if image already exists on host
+	_, _, err = dockerClient.ImageInspectWithRaw(ctx, imageName)
+	if err == nil {
+		// image already exists
+		return imageName, nil
+	} else if strings.Contains(err.Error(), "No such image") {
```

Comment:
This is potentially brittle. If docker changes its error message, this won't work anymore. It might make sense to just outdent all of this and get rid of the else block.

Reply:
To document our convo: I agree this isn't ideal, but ImageInspectWithRaw could fail for other reasons and it's important to catch both "no such image" errors and all other errrors. The Docker Engine API returns 404 and 500 error codes in such cases, but the vendored functions sadly don't pass this on to Evergreen.